### PR TITLE
feat(makefile): add support for makefiles

### PR DESCRIPTION
**Pull request by eganjs**
*Issue #, if available:*
n/a

*Description of changes:*
Add class to generate simple Makefiles

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/makefile.ts [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+  isPhony?: boolean;
+}
+
+export interface MakefileOptions extends FileBaseOptions {
+  all?: string[];
+  rules: Rule[];
+}
+
+export class Makefile extends FileBase {
+  readonly all?: string[];
```
Comment (eladb) [inline src/makefile.ts:17]: Can we just model this as a built-in rule (defined in the ctor).

**Review by eladb — COMMENTED** [review]
This is asking for some high level APIs, but those can be added later. For example, it would be nice if the prerequisites are not just strings but rather references to either existing files or artifacts of a rule. Things like that.

File: src/makefile.ts [hunk 2]
```diff
@@ -0,0 +1,55 @@
+import { FileBase, FileBaseOptions, IResolver } from './file';
+import { Project } from './project';
+
+export interface Rule {
+  targets: string[];
+  prerequisites?: string[];
```
[the remark this answers is not in the document]
Comment (eganjs) [reply src/makefile.ts:6]: I've based my terminology on this page from the Make manual https://www.gnu.org/software/make/manual/html_node/Rule-Introduction.html

File: src/makefile.ts (hunk 1 above)
Comment (eganjs) [reply src/makefile.ts:17]: I'm not sure I understand, what does "ctor" refer to?

[the remark this answers is not in the document]
Comment (eganjs) [reply src/makefile.ts]: done

File: src/makefile.ts [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+  recipe?: string[];
+  isPhony?: boolean;
+}
+
+export interface MakefileOptions extends FileBaseOptions {
+  all?: string[];
+  rules: Rule[];
+}
+
+export class Makefile extends FileBase {
```
[the remark this answers is not in the document]
Comment (eganjs) [reply src/makefile.ts:51]: done

File: src/makefile.ts [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+  readonly all?: string[];
+  readonly rules: Rule[];
```
[the remark this answers is not in the document]
Comment (eganjs) [reply src/makefile.ts]: done

File: src/makefile.ts [hunk 5]
```diff
@@ -0,0 +1,55 @@
+import { FileBase, FileBaseOptions, IResolver } from './file';
```
[the remark this answers is not in the document]
Comment (eganjs) [reply src/makefile.ts:1]: done

File: src/makefile.ts [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+import { Project } from './project';
+
+export interface Rule {
+  targets: string[];
+  prerequisites?: string[];
+  recipe?: string[];
+  isPhony?: boolean;
```
[the remark this answers is not in the document]
Comment (eganjs) [reply src/makefile.ts]: done

File: src/makefile.ts [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+
+  /**
+   * Commands that are run (using prerequisites as inputs) to create a target.
+   */
+  readonly recipe?: string[];
+
+  /**
+   * Marks whether the target is phony.
+   *
+   * False by default.
```
Comment (eladb) [inline src/makefile.ts:28]:
```suggestion
   * @default false
```

File: src/makefile.ts [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+  readonly targets: string[];
+
+  /**
+   * Files that are used as inputs to create a target.
+   */
+  readonly prerequisites?: string[];
+
+  /**
+   * Commands that are run (using prerequisites as inputs) to create a target.
+   */
```
Comment (eladb) [inline src/makefile.ts:22]: "@default" is required for all optional fields. In this case, you can do something like this:

```
@default []
```

File: src/makefile.ts [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+
+  constructor(project: Project, filePath: string, options: MakefileOptions) {
+    super(project, filePath, options);
+
+    if (!options.rules) {
+      throw new Error('"rules" cannot be undefined');
+    }
+
+    this.all = options.all;
+    this.rules = options.rules;
```
Comment (eladb) [inline src/makefile.ts:71]: You can just implement `all` through `rules` I believe:

```suggestion
    const all = { phony: true, targets: ['all'], prerequisites: options.all };
    this.rules = [
       ...options.all ? all : [],
       ...options.rules,
    ];
```

File: src/makefile.ts [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+  }
+
```
Comment (eladb) [inline src/makefile.ts:73]: Would be nice to also add imperative API:

```ts
makefile.addRule(rule);
makefile.addAll(targetName);
```

File: src/makefile.ts (hunk 8 above)
Comment (eganjs) [reply src/makefile.ts:26]: done

File: src/makefile.ts (hunk 7 above)
Comment (eganjs) [reply src/makefile.ts]: done

File: src/makefile.ts (hunk 10 above)
Comment (eganjs) [reply src/makefile.ts:129]: Done. I've used a builder like pattern so the calls can be chained. e.g.
```
makefile
  .addRule(rule1)
  .addRule(rule2);
```

File: src/makefile.ts (hunk 9 above)
Comment (eganjs) [reply src/makefile.ts]: done

Comment (eganjs) [reply src/makefile.ts]: Also it turns out
```
.PHONY: all
all:
```
is a valid Makefile, so I've rewritten it to always add the all rule

**Review by eladb — APPROVED** [review]

**Pull request merged**